File: solutions/sheet4/intersect.cpp
#include <iostream>
#include <iomanip>
#include <vector>
#include <cmath>
#include <algorithm>
#include <stdlib.h>
#include <random>

using namespace std;

typedef float coord_type;

struct Point
{
    coord_type x;
    coord_type y;

    Point(coord_type xv, coord_type yv) : x(xv), y(yv) {}
    Point() : x(0.), y(0.) {}
};

typedef vector<Point> Polygon;


int orientation(const Point p1, const Point p2, const Point p3)
{
    coord_type val = (p2.y - p1.y) * (p3.x - p2.x) - (p2.x - p1.x) * (p3.y - p2.y);
    return val > 0 ? 1 : -1;
}

bool linesIntersect(const Point p1, const Point p2, const Point q1, const Point q2)
{
    int o1 = orientation(p1, p2, q1);
    int o2 = orientation(p1, p2, q2);
    if(o1 != o2)
    {
        int o3 = orientation(q1, q2, p1);
        int o4 = orientation(q1, q2, p2);
        return o3 != o4;
    }

    return 0;
}


int pointInPolygon(const Point p1, const Polygon &polygon)
{
    int c = 0;
    int n = polygon.size();
    Point p2 = Point(11, p1.y);
    for(int i = 0; i <= n; i++) {
        Point q1 = polygon[i];
        Point q2 = polygon[(i+1)%n];
        if(linesIntersect(q1,q2,p1,p2)) { c++; }
    }

    return c % 2 == 1;
}


bool pointInAllPolygons(const Point p, const vector<Polygon> &polygons)
{
    for(auto polygon : polygons)
    {
        if(!pointInPolygon(p, polygon)) {return false;}
    }
    return true;
}

coord_type polygonVolume(const Polygon polygon)
{
    coord_type sum = 0;
    int n = polygon.size();
    for(int i = 0; i <= n; i++)
    {
        Point p1 = polygon[i];
        Point p2 = polygon[(i+1)%n];
        sum += (p1.x + p2.x) * (p1.y - p2.y);
    }
    return 0.5 * sum;
}


int main()
{
    int n;
	cin >> n;

    vector<Polygon> polygons;
    coord_type minX, minY;
    coord_type maxX, maxY;
    minX = minY = 0;
    maxX = maxY = 10;

    // read polygons and find smallest common bounding box
    for(int i = 0; i < n; i++)
    {
        int ki;
        cin >> ki;

        coord_type pminX, pminY;
        coord_type pmaxX, pmaxY;
        pminX = pminY = 10;
        pmaxX = pmaxY = 0;

        vector<Point> polygon;
        for(int j=0; j < ki; j++)
        {
            coord_type vx, vy;
            cin >> vx;
            cin >> vy;
            polygon.push_back(Point(vx,vy));
            pminX = min(pminX, vx);
            pminY = min(pminY, vy);
            pmaxX = max(pmaxX, vx);
            pmaxY = max(pmaxY, vy);
        }
        polygons.push_back(polygon);
        minX = max(minX, pminX);
        minY = max(minY, pminY);
        maxX = min(maxX, pmaxX);
        maxY = min(maxY, pmaxY);
    }

    // sort polygons by size to improve early stopping
    sort(polygons.begin(), polygons.end(), [](Polygon p1, Polygon p2) {return polygonVolume(p1) > polygonVolume(p2); });

    int numSamples = 800000;
    int numHits = 0;

    coord_type bbArea = (maxX-minX) * (maxY-minY);

    // set up rng
    random_device rd;
    mt19937_64 engine(rd());
    uniform_real_distribution<coord_type> distX(minX,maxX);
    uniform_real_distribution<coord_type> distY(minY,maxY);

    // approximate area with monte carlo
    for(int i = 0; i < numSamples; i++)
    {
        Point s(distX(engine), distY(engine));
        if(pointInAllPolygons(s, polygons)) {numHits++;}
    }

    coord_type area = bbArea * (coord_type(numHits) / coord_type(numSamples));
    cout << fixed << setprecision(2) << area << endl;
    return 0;
}

File: solutions/sheet4/intersect.py
import sys
from random import random
from time import time

stime = time()

n = int(sys.stdin.readline())
lines = [[int(v) for v in line.split(" ")] for line in sys.stdin]
polygons = [line[1:] for line in lines]
polygons = [[(x, y) for x, y in zip(p[::2], p[1::2])] for p in polygons]


def orientation(p1, p2, p3):
    val = (p2[1] - p1[1]) * (p3[0] - p2[0]) - (p2[0] - p1[0]) * (p3[1] - p2[1])
    return 1 if val > 0 else -1


def lines_intersect(p1, p2, q1, q2):
    o1 = orientation(p1, p2, q1)
    o2 = orientation(p1, p2, q2)
    if o1 != o2:
        o3 = orientation(q1, q2, p1)
        o4 = orientation(q1, q2, p2)
        return o3 != o4

    return False


def point_in_polygon(p1, polygon):
    i = 0
    p2 = (11, p1[1])
    for q1, q2 in zip(polygon, polygon[1:] + [polygon[0]]):
        if lines_intersect(q1, q2, p1, p2):
            i += 1

    return i % 2 == 1


def get_sample_in_bb(min_x, min_y, max_x, max_y):
    return (max_x - min_x) * random() + min_x, (max_y - min_y) * random() + min_y


def polygon_volume(points):
    return 0.5 * sum((p1[0] + p2[0]) * (p1[1] - p2[1]) for p1, p2 in zip(points, points[1:] + [points[0]]))


# find minimal bounding box
min_x = min_y = 0
max_x = max_y = 10

for polygon in polygons:
    cur_min_x = min(p[0] for p in polygon)
    cur_min_y = min(p[1] for p in polygon)
    cur_max_x = max(p[0] for p in polygon)
    cur_max_y = max(p[1] for p in polygon)
    min_x = max(min_x, cur_min_x)
    min_y = max(min_y, cur_min_y)
    max_x = min(max_x, cur_max_x)
    max_y = min(max_y, cur_max_y)


polygons.sort(key=polygon_volume)

bb_area = (max_x - min_x) * (max_y - min_y)
num_samples = 0
num_hits = 0

while time() - stime < 0.9:
    num_samples += 1
    s = get_sample_in_bb(min_x, min_y, max_x, max_y)
    if all(point_in_polygon(s, poly) for poly in polygons):
        num_hits += 1

area = bb_area * num_hits / num_samples

print("{:.2f}".format(area))
